src/server/src/main/java/io/cassandrareaper/service/Heart.java

```diff
@@ -0,0 +1,188 @@
+
+  private final AtomicLong lastBeat = new AtomicLong(System.currentTimeMillis() - TimeUnit.HOURS.toMillis(1));
+  private final ForkJoinPool forkJoinPool = new ForkJoinPool(64);
+  private final long maxBeatFrequencyMillis;
+  private final AtomicBoolean updatingNodeMetrics = new AtomicBoolean(false);
+
+  private Heart(long maxBeatFrequency) {
+    this.maxBeatFrequencyMillis = maxBeatFrequency;
+  }
+
+  static Heart create() {
+    return new Heart(DEFAULT_MAX_FREQUENCY);
+  }
+
+  @VisibleForTesting
+  static Heart create(long maxBeatFrequencyMillis) {
+    return new Heart(maxBeatFrequencyMillis);
+  }
+
+  synchronized void beat(AppContext context) {
+    if (context.storage instanceof IDistributedStorage
+        && lastBeat.get() + maxBeatFrequencyMillis < System.currentTimeMillis()) {
+
+      lastBeat.set(System.currentTimeMillis());
+      ((IDistributedStorage) context.storage).saveHeartbeat();
+
+      if (ReaperApplicationConfiguration.DatacenterAvailability.EACH == context.config.getDatacenterAvailability()) {
+        updateAllReachableEmptyNodeMetrics(context);
+      }
+    }
+  }
+
+  AtomicBoolean isCurrentlyUpdatingNodeMetrics() {
+    return new AtomicBoolean(updatingNodeMetrics.get());
+  }
+
+  @Override
+  public void close() {
+    try {
+      forkJoinPool.shutdown();
+      forkJoinPool.awaitTermination(10, TimeUnit.SECONDS);
+    } catch (InterruptedException ignore) {
+    } finally {
+      forkJoinPool.shutdownNow();
+    }
+  }
+
+  private void updateAllReachableEmptyNodeMetrics(AppContext context) {
+    Preconditions.checkArgument(context.storage instanceof IDistributedStorage);
+    IDistributedStorage storage = ((IDistributedStorage) context.storage);
+    registerGuages(context);
+
+    if (!updatingNodeMetrics.getAndSet(true)) {
+      HostConnectionCounters hostConnectionCounters = context.jmxConnectionFactory.getHostConnectionCounters();
+      int jmxTimeoutSeconds = context.config.getJmxConnectionTimeoutInSeconds();
+
+      forkJoinPool.submit(() -> {
+        try (Timer.Context t0 = timer(context, "metrics", "all")) {
+
+          Set<String> metrics = storage.getNodeMetrics().stream()
+              .map(nodeMetric -> nodeMetric.getHostAddress())
+              .collect(Collectors.toSet());
+
+          forkJoinPool.submit(() -> {
+            context.storage.getClusters().parallelStream().forEach(cluster -> {
+
+              try (Timer.Context t1 = timer(context, "metrics", cluster.getName())) {
+                Set<String> updatedNodes = Sets.newConcurrentHashSet(metrics);
+
+                try (JmxProxy seedProxy = context.jmxConnectionFactory.connectAny(cluster, jmxTimeoutSeconds)) {
+
+                  seedProxy.getLiveNodes().parallelStream()
+                      .filter(node -> 0 <= hostConnectionCounters.getSuccessfulConnections(node))
+                      .filter(node -> updatedNodes.add(node))
+                      .forEach(node -> {
+
+                        try (
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Extract this nested try block into a separate method. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1141)